README.md
# formula

Workbench module that runs Excel or Python formula between columns and stores result in a new column

# Development

Do the Development Dance:

1. Run `tox` and make sure it works
2. Write a failing test to `tests/`
3. Run `tox` to test it fails
4. Fix the code until `tox` passes
5. Submit a pull request

# Deployment

1. Write an entry to `CHANGELOG.md`
2. Run `tox` one last time
3. `git push`


CHANGELOG.md
2021-05-31.01
-------------

* Fix crash related to `Date` handling.

2021-04-23.01
-------------

* Handle `Date` the same way we handle `Timestamp`. In Excel, all dates are
  Numbers. When the result is a Number, we don't interpret it as a Workbench
  Date -- same as with timestamp, users see the unformatted Excel Numbers.)

2020-09-15.02
-------------

* Return '#VALUE!' as Text. Previously, Workbench would notify that the step
  crashed if it returned '#VALUE!'

2020-09-15.01
-------------

* Convert True/False formula output into "True" and "False" (Text). Previously,
  Workbench would notify that the step crashed.


test_formula.py
import unittest
from typing import Any, Dict
import numpy as np
import pandas as pd
from pandas.testing import assert_frame_equal, assert_series_equal
import formula
from formula import build_globals_for_eval, sanitize_series
from cjwmodule.testing.i18n import i18n_message


class MockParams:
    @staticmethod
    def factory(**kwargs):
        """Build a MockParams factory with default values.

        Usage:

            P = MockParams.factory(foo=3)
            params = P(bar=2)  # {'foo': 3, 'bar': 2}
        """
        return lambda **d: {**kwargs, **d}


P = MockParams.factory(
    syntax="excel", out_column="R", formula_excel="", all_rows=False, formula_python=""
)


class MigrateParamsTests(unittest.TestCase):
    def test_v0_excel(self):
        result = formula.migrate_params(
            {
                "syntax": 0,
                "out_column": "R",
                "formula_excel": "=A1",
                "formula_python": "A",
                "all_rows": True,
            }
        )
        self.assertEqual(
            result,
            {
                "syntax": "excel",
                "out_column": "R",
                "formula_excel": "=A1",
                "formula_python": "A",
                "all_rows": True,
            },
        )

    def test_v0_python(self):
        result = formula.migrate_params(
            {
                "syntax": 1,
                "out_column": "R",
                "formula_excel": "=A1",
                "formula_python": "A",
                "all_rows": True,
            }
        )
        self.assertEqual(
            result,
            {
                "syntax": "python",
                "out_column": "R",
                "formula_excel": "=A1",
                "formula_python": "A",
                "all_rows": True,
            },
        )

    def test_v1(self):
        result = formula.migrate_params(
            {
                "syntax": "python",
                "out_column": "R",
                "formula_excel": "=A1",
                "formula_python": "A",
                "all_rows": True,
            }
        )
        self.assertEqual(
            result,
            {
                "syntax": "python",
                "out_column": "R",
                "formula_excel": "=A1",
                "formula_python": "A",
                "all_rows": True,
            },
        )


class SanitizeSeriesFrameTest(unittest.TestCase):
    def test_mixed_to_string_keeps_nan(self):
        # check that sanitizing a non-string column with missing data produces
        # empty cells, not 'nan' strings
        # https://www.pivotaltracker.com/story/show/154619564
        series = pd.Series([1.0, "str", np.nan, ""])  # mixed
        result = sanitize_series(series)
        assert_series_equal(result, pd.Series(["1.0", "str", np.nan, ""]))

    def test_mixed_to_string_allows_custom_types(self):
        class Obj:
            def __str__(self):
                return "x"

        series = pd.Series([Obj(), Obj()])
        result = sanitize_series(series)
        expected = pd.Series(["x", "x"])
        assert_series_equal(result, expected)

    def test_categories_to_string_allows_custom_category_types(self):
        class Obj:
            def __init__(self, value):
                self.value = value

            def __str__(self):
                return self.value

        series = pd.Series([Obj("a"), Obj("b"), Obj("a"), "a", "y"], dtype="category")
        result = sanitize_series(series)
        expected = pd.Series(["a", "b", "a", "a", "y"], dtype="category")
        assert_series_equal(result, expected)

    def test_categories_to_string_allows_abnormal_index(self):
        class Obj:
            def __init__(self, value):
                self.value = value

            def __str__(self):
                return self.value

        # Slicing a Series means the category list remains complete, even
        # though some categories aren't used. In this example, `series` has an
        # Obj('a') category, even though the value doesn't appear anywhere in
        # the dataframe. (This is because slicing creates a numpy "view", not a
        # copy of the original array of codes.)
        #
        # Sanitize's output shouldn't include any categories that aren't
        # visible. (The data in memory should not be a "view".)
        #
        # Also, sanitize_series() should reset the index.
        series = pd.Series([Obj("a"), Obj("b"), "c", "b"], dtype="category")[1:]
        result = sanitize_series(series)
        expected = pd.Series(["b", "c", "b"], dtype="category")
        assert_series_equal(result, expected)
        # to reiterate: 'result' has no category that looks like 'a'.
        self.assertEqual(sorted(result.cat.categories.tolist()), ["b", "c"])

    def test_lists_and_dicts(self):
        series = pd.Series([[5, 6, 7], {"a": "b"}])
        result = sanitize_series(series)
        expected = pd.Series(["[5, 6, 7]", "{'a': 'b'}"])
        assert_series_equal(result, expected)

    def test_reset_index(self):
        # should always come out with row numbers contiguous from zero
        series = pd.Series([1, 2, 3])[1:]
        result = sanitize_series(series)
        assert_series_equal(result, pd.Series([2, 3]))  # index is [0,1]

    def test_cast_int_category_to_int(self):
        series = pd.Series([1, 2], dtype="category")
        result = sanitize_series(series)
        assert_series_equal(result, pd.Series([1, 2]))

    def test_cast_mixed_category_to_str(self):
        series = pd.Series([1, "2"], dtype="category")
        result = sanitize_series(series)
        expected = pd.Series(["1", "2"], dtype="category")
        assert_series_equal(result, expected)

    def test_remove_unused_categories(self):
        series = pd.Series(
            ["a", "b"],
            # extraneous value
            dtype=pd.api.types.CategoricalDtype(["a", "b", "c"]),
        )
        result = sanitize_series(series)
        expected = pd.Series(["a", "b"], dtype="category")
        assert_series_equal(result, expected)


class FormulaTests(unittest.TestCase):
    def _test(
        self,
        table: pd.DataFrame,
        params: Dict[str, Any] = {},
        expected_table: pd.DataFrame = pd.DataFrame(),
        expected_error: str = "",
    ):
        result = formula.render(table, P(**params))
        if expected_error:
            self.assertEqual(result, expected_error)
        else:
            self.assertIsInstance(result, pd.DataFrame)
            assert_frame_equal(result, expected_table)

    def test_python_formula_int_output(self):
        self._test(
            pd.DataFrame({"A": [10, 20]}),
            {"syntax": "python", "formula_python": "A*2"},
            pd.DataFrame({"A": [10, 20], "R": [20, 40]}),
        )

    def test_python_formula_str_output(self):
        self._test(
            pd.DataFrame({"A": [10, 20]}),
            {"syntax": "python", "formula_python": 'str(A) + "x"'},
            pd.DataFrame({"A": [10, 20], "R": ["10x", "20x"]}),
        )

    def test_python_formula_empty_output_pval_makes_result(self):
        # empty out_column defaults to 'result'
        self._test(
            pd.DataFrame({"A": [1]}),
            {"syntax": "python", "formula_python": "A*2", "out_column": ""},
            pd.DataFrame({"A": [1], "result": [2]}),
        )

    def test_python_formula_missing_colname_makes_error(self):
        # formula with missing column name should error
        self._test(
            pd.DataFrame({"A": [1]}),
            {"syntax": "python", "formula_python": "B*2"},
            expected_error="name 'B' is not defined",
        )

    def test_python_formula_cast_nonsane_output(self):
        self._test(
            pd.DataFrame({"A": [1, 2]}),
            {"syntax": "python", "formula_python": "[A]"},
            pd.DataFrame({"A": [1, 2], "R": ["[1]", "[2]"]}),
        )

    def test_python_spaces_to_underscores(self):
        # column names with spaces should be referenced with underscores in the
        # formula
        self._test(
            pd.DataFrame({"A b": [1, 2]}),
            {"syntax": "python", "formula_python": "A_b*2"},
            pd.DataFrame({"A b": [1, 2], "R": [2, 4]}),
        )

    def test_excel_formula_no_output_col_name(self):
        # if no output column name specified, store to a column named 'result'
        self._test(
            pd.DataFrame({"A": [1, 2]}),
            {"formula_excel": "=A1*2", "all_rows": True, "out_column": ""},
            pd.DataFrame({"A": [1, 2], "result": [2.0, 4.0]}),
        )

    def test_excel_date(self):
        self._test(
            pd.DataFrame({"A": [1, 2, 3]}),
            {
                "formula_excel": "=DATE(2019, 6, A1)",
                "all_rows": True,
                "out_column": "X",
            },
            pd.DataFrame(
                {
                    "A": [1, 2, 3],
                    # Excel dates are integers. TODO let user specify an output
                    # format, so they get useful behavior.
                    "X": [43617, 43618, 43619],
                }
            ),
        )

    def test_excel_bool(self):
        self._test(
            pd.DataFrame({"A": [1, 2, None]}),
            {
                "formula_excel": "=A1=1",
                "all_rows": True,
                "out_column": "X",
            },
            # "None=1" evaluates to False, in Excel-land
            pd.DataFrame({"A": [1, 2, None], "X": ["True", "False", "False"]}),
        )

    def test_excel_bool_mixed_type(self):
        self._test(
            pd.DataFrame({"A": [1, 2, None]}),
            {
                "formula_excel": "=IF(A1=1, TRUE, 3)",
                "all_rows": True,
                "out_column": "X",
            },
            # "None=1" evaluates to False, in Excel-land
            pd.DataFrame({"A": [1, 2, None], "X": ["True", "3", "3"]}),
        )

    def test_excel_timestamp_add(self):
        # This is really tricky, because of a bugs in Lotus-1-2-3 in the 1980s.
        # https://docs.microsoft.com/en-gb/office/troubleshoot/excel/wrongly-assumes-1900-is-leap-year
        #
        # Let's follow the behavior of LibreOffice and Google Sheets: say the
        # date system starts at 1899-12-30 instead of 1899-12-31. Test:
        #
        # 1. in column A, enter some dates. (1900-01-01 becomes 2.0!)
        # 2. in column B, enter "=A1" and format as Number
        # 3. in column C, enter "=A1+3"
        # 4. in column D, enter "=A1+3" (again) and format as Datetime
        dates = pd.Series(
            ["1900-01-01", "1900-02-27", "2020-03-01"], dtype="datetime64[ns]"
        )
        self._test(
            pd.DataFrame({"A": dates}),
            {
                "formula_excel": "=A1 + 3",
                "all_rows": True,
                "out_column": "X",
            },
            pd.DataFrame(
                {
                    "A": dates,
                    # Excel dates are integers. TODO let user specify an output
                    # format, so they get useful behavior.
                    "X": [5.0, 62.0, 43894.0],
                }
            ),
        )

    def test_excel_date_add(self):
        # Look to `test_excel_timestamp_add` for canonical test
        dates = pd.Series(["1900-01-01", "1900-02-27", "2020-03-01"], dtype="period[D]")
        self._test(
            pd.DataFrame({"A": dates}),
            {
                "formula_excel": "=A1 + 3",
                "all_rows": True,
                "out_column": "X",
            },
            pd.DataFrame(
                {
                    "A": dates,
                    # Excel dates are integers. TODO let user specify an output
                    # format, so they get useful behavior.
                    "X": [5.0, 62.0, 43894.0],
                }
            ),
        )

    def test_excel_all_rows_function_not_implemented(self):
        self._test(
            pd.DataFrame({"A": [1, 2, 3]}),
            {
                "formula_excel": "=DATEX(2019, 6, A1)",
                "all_rows": True,
                "out_column": "X",
            },
            expected_error=i18n_message(
                "excel.functionNotImplemented", {"name": "DATEX"}
            ),
        )

    # --- Formulas which write to all rows ---
    def test_excel_all_rows_single_column(self):
        self._test(
            pd.DataFrame({"A": [1, 2]}),
            {"formula_excel": "=A1*2", "all_rows": True},
            pd.DataFrame({"A": [1, 2], "R": [2.0, 4.0]}),
        )

    def test_excel_all_rows_column_range(self):
        self._test(
            pd.DataFrame({"A": [1, 2], "B": [2, 3], "C": [3, 4]}),
            {"formula_excel": "=SUM(A1:C1)", "all_rows": True},
            pd.DataFrame({"A": [1, 2], "B": [2, 3], "C": [3, 4], "R": [6, 9]}),
        )

    def test_excel_text_formula(self):
        self._test(
            pd.DataFrame({"A": ["foo", "bar"]}),
            {"formula_excel": "=LEFT(A1, 2)", "all_rows": True},
            pd.DataFrame({"A": ["foo", "bar"], "R": ["fo", "ba"]}),
        )

    # --- Formulas which write only to a single row ---
    def test_excel_divide_two_rows(self):
        self._test(
            pd.DataFrame({"A": [1, 2]}),
            {"formula_excel": "=A1/A2", "all_rows": False},
            pd.DataFrame({"A": [1, 2], "R": [0.5, np.nan]}),
        )

    def test_excel_add_two_columns(self):
        self._test(
            pd.DataFrame({"A": [1, 2], "B": [2, 3]}),
            {"formula_excel": "=A1+B1", "all_rows": False},
            pd.DataFrame({"A": [1, 2], "B": [2, 3], "R": [3, np.nan]}),
        )

    def test_excel_function_not_implemented(self):
        self._test(
            pd.DataFrame({"A": [1, 2, 3]}),
            {
                "formula_excel": "=DATEX(2019, 6, A1)",
                "all_rows": False,
                "out_column": "X",
            },
            expected_error=i18n_message(
                "excel.functionNotImplemented", {"name": "DATEX"}
            ),
        )

    def test_excel_sum_column(self):
        self._test(
            pd.DataFrame({"A": [1, 2, 3]}),
            {"formula_excel": "=SUM(A2:A3)", "all_rows": False},
            pd.DataFrame({"A": [1, 2, 3], "R": [5, np.nan, np.nan]}),
        )

    def test_excel_error_missing_row_number(self):
        self._test(
            pd.DataFrame({"A": [1, 2]}),
            {"formula_excel": "=A*2", "all_rows": True},
            expected_error=i18n_message("excel.badCellReference", {"token": "A"}),
        )

    def test_excel_error_missing_row_number_in_range(self):
        self._test(
            pd.DataFrame({"A": [1, 2], "B": [2, 3]}),
            {"formula_excel": "=SUM(A:B)", "all_rows": True},
            expected_error=i18n_message("excel.formulaFirstRowReference"),
        )

    def test_excel_error_reference_row_other_than_1_with_all_rows(self):
        self._test(
            pd.DataFrame({"A": [1, 2]}),
            {"formula_excel": "=A2*2", "all_rows": True},
            expected_error=i18n_message("excel.formulaFirstRowReference"),
        )

    def test_excel_error_syntax(self):
        self._test(
            pd.DataFrame({"A": [1, 2]}),
            {"formula_excel": "=SUM B>", "all_rows": False},
            expected_error=i18n_message(
                # The "%s" is built in to the formulas module. TODO file bugrep
                "excel.invalidFormula",
                {"error": "Not a valid formula:\n%s"},
            ),
        )

    def test_excel_error_out_of_range_columns(self):
        self._test(
            pd.DataFrame({"A": [1, 2]}),
            {"formula_excel": "=SUM(A1:B1)", "all_rows": True},
            expected_error=i18n_message(
                "excel.all_rows.badColumnRef", {"ref": "A1:B1"}
            ),
        )

    def test_excel_error_out_of_range_rows(self):
        self._test(
            pd.DataFrame({"A": [1, 2]}),
            {"formula_excel": "=SUM(A1:A3)", "all_rows": False},
            # Just ignore missing rows
            pd.DataFrame({"A": [1, 2], "R": [3, np.nan]}),
        )

    def test_excel_error_row_0(self):
        self._test(
            pd.DataFrame({"A": [1, 2]}),
            {"formula_excel": "=A0*2", "all_rows": False},
            expected_error=i18n_message(
                "excel.one_row.invalidCellRange", {"token": "A0"}
            ),
        )

    def test_excel_sanitize_output(self):
        self._test(
            pd.DataFrame({"A": [1, 2]}),
            {"formula_excel": '=IF(A1=1, 1, "x")', "all_rows": True},
            pd.DataFrame({"A": [1, 2], "R": ["1", "x"]}),
        )

    def test_excel_value_error(self):
        # https://www.pivotaltracker.com/story/show/168909000
        self._test(
            pd.DataFrame({"A": ["1", "x", "3"]}),
            {"formula_excel": "=SUM(A1:A3)", "all_rows": False, "out_column": "X"},
            pd.DataFrame({"A": ["1", "x", "3"], "X": ["#VALUE!", None, None]}),
        )


class SafeExecTest(unittest.TestCase):
    def exec_code(self, code):
        built_globals = build_globals_for_eval()
        inner_locals = {}
        exec(code, built_globals, inner_locals)
        return inner_locals

    def test_builtin_functions(self):
        env = self.exec_code(
            """
ret = sorted(list([1, 2, sum([3, 4])]))
"""
        )
        self.assertEqual(env["ret"], [1, 2, 7])


formula.py
import builtins
import itertools
import math
from typing import Any, Dict

import numpy as np
import pandas as pd
from cjwmodule import i18n
from formulas import Parser
from pandas.api.types import is_datetime64_dtype, is_numeric_dtype
from schedula import DispatcherError, Token


def autocast_series_dtype(series: pd.Series) -> pd.Series:
    """Cast any sane Series to str/category[str]/number/datetime.

    This is appropriate when parsing CSV data or Excel data. It _seems_
    appropriate when a search-and-replace produces numeric columns like
    '$1.32' => '1.32' ... but perhaps that's only appropriate in very-specific
    cases.

    The input must be "sane": if the dtype is object or category, se assume
    _every value_ is str (or null).

    If the series is all-null, do nothing.

    Avoid spurious calls to this function: it's expensive.

    TODO handle dates.
    """
    if series.dtype == bool:
        # Handle Excel formula: '=TRUE'
        #
        # We capitalize Pythonic, because A) sometimes the formula is in Python;
        # and B) the `series.astype(str)` below is hard to customize
        return series.replace({True: "True", False: "False"})
    elif series.dtype == object:
        nulls = series.isnull()
        if (nulls | (series == "")).all():
            return series
        try:
            # If it all looks like numbers (like in a CSV), cast to number.
            return pd.to_numeric(series)
        except (ValueError, TypeError):
            # Otherwise, we want all-string. Is that what we already have?
            #
            # Handles Excel formula: =IF(A1=1, 3, "Hi")
            array = series[~nulls].array
            if any(type(x) != str for x in array):
                series = series.astype(str)
                series[nulls] = None
            return series
    elif hasattr(series, "cat"):
        # Categorical series. Try to infer type of series.
        #
        # Assume categories are all str: after all, we're assuming the input is
        # "sane" and "sane" means only str categories are valid.
        if (series.isnull() | (series == "")).all():
            return series
        try:
            return pd.to_numeric(series)
        except (ValueError, TypeError):
            # We don't cast categories to str here -- because we have no
            # callers that would create categories that aren't all-str. If we
            # ever do, this is where we should do the casting.
            return series
    else:
        assert is_numeric_dtype(series) or is_datetime64_dtype(series)
        return series


class UserVisibleError(Exception):
    """Has an `i18n.I18nMessage` as its first argument"""

    @property
    def i18n_message(self):
        return self.args[0]


def build_builtins_for_eval() -> Dict[str, Any]:
    """
    Build a __builtins__ for use in custom code.

    Call ``exec(code, {'__builtins__': retval}, {})`` to use it.
    """
    # Start with _this_ module's __builtins__
    eval_builtins = dict(builtins.__dict__)

    # Disable "dangerous" builtins.
    #
    # This doesn't increase security: it just helps module authors.
    def disable_func(name):
        def _disabled(*args, **kwargs):
            raise UserVisibleError(
                i18n.trans(
                    "python.disabledFunction",
                    "{name} is disabled",
                    {"name": "builtins.%s" % name},
                )
            )

        return _disabled

    to_disable = ["__import__", "breakpoint", "compile", "eval", "exec", "open"]
    for name in to_disable:
        eval_builtins[name] = disable_func(name)

    return eval_builtins


def build_globals_for_eval() -> Dict[str, Any]:
    """Build a __globals__ for use in custom code."""
    eval_builtins = build_builtins_for_eval()
    return {"__builtins__": eval_builtins, "math": math, "np": np, "pd": pd}


def sanitize_series(series: pd.Series) -> pd.Series:
    """Enforce type rules on input pandas `Series.values`.

    The return value is anything that can be passed to the `pandas.Series()`
    constructor.

    Specific fixes:

    * Make sure categories have no excess values.
    * Convert numeric categories to
    * Convert unsupported dtypes to string.
    * Reindex so row numbers are contiguous.
    """
    series.reset_index(drop=True, inplace=True)
    if hasattr(series, "cat"):
        series.cat.remove_unused_categories(inplace=True)

        categories = series.cat.categories
        if pd.api.types.is_numeric_dtype(categories.values):
            # Un-categorize: make array of int/float
            return pd.to_numeric(series)
        elif (
            categories.dtype != object
            or pd.api.types.infer_dtype(categories.values, skipna=True) != "string"
        ):
            # Map from non-Strings to Strings
            #
            # 1. map the _codes_ to unique _codes_
            mapping = pd.Categorical(categories.astype(str))
            values = pd.Categorical(series.cat.codes[mapping.codes])
            # 2. give them names
            values.rename_categories(mapping.categories, inplace=True)
            series = pd.Series(values)

        return series
    elif is_numeric_dtype(series.dtype):
        return series
    elif is_datetime64_dtype(series.dtype):
        return series
    else:
        # Force it to be a str column: every object is either str or np.nan
        ret = series.astype(str)
        ret[pd.isna(series)] = np.nan
        return ret


def python_formula(table, formula):
    # spaces to underscores in column names
    colnames = [x.replace(" ", "_") for x in table.columns]

    code = compile(formula, "<string>", "eval")
    custom_code_globals = build_globals_for_eval()

    # Much experimentation went into the form of this loop for good
    # performance.
    # Note we don't use iterrows or any pandas indexing, and construct the
    # values dict ourselves
    newcol = pd.Series(list(itertools.repeat(None, len(table))))
    for i, row in enumerate(table.values):
        newcol[i] = eval(code, custom_code_globals, dict(zip(colnames, row)))

    newcol = autocast_series_dtype(sanitize_series(newcol))

    return newcol


def flatten_single_element_lists(x):
    """Return `x[0]` if `x` is a list, otherwise `x`."""
    if isinstance(x, list) and len(x) == 1:
        return x[0]
    else:
        return x


def eval_excel(code, args):
    """Return result of running Excel code with args.

    Raise UserVisibleError if a function is unimplemented.
    """
    try:
        ret = code(*args)
    except DispatcherError as err:
        if isinstance(err.args[2], NotImplementedError):
            raise UserVisibleError(
                i18n.trans(
                    "excel.functionNotImplemented",
                    "Function {name} not implemented",
                    {"name": err.args[1]},
                )
            )
        else:
            raise
    if isinstance(ret, np.ndarray):
        return ret.item()
    else:
        return ret


def eval_excel_one_row(code, table):

    # Generate a list of input table values for each range in the expression
    formula_args = []
    for token, obj in code.inputs.items():
        if obj is None:
            raise UserVisibleError(
                i18n.trans(
                    "excel.one_row.invalidCellRange",
                    "Invalid cell range: {token}",
                    {"token": token},
                )
            )
        ranges = obj.ranges
        if len(ranges) != 1:
            # ...not sure what input would get us here
            raise UserVisibleError(
                i18n.trans(
                    "excel.one_row.cellRangeNotRectangular",
                    "Excel range must be a rectangular block of values",
                )
            )
        range = ranges[0]

        # Unpack start/end row/col
        r1 = int(range["r1"]) - 1
        r2 = int(range["r2"])
        c1 = int(range["n1"]) - 1
        c2 = int(range["n2"])

        nrows, ncols = table.shape
        # allow r2 > nrows: users use it to say SUM(A1:A99999)
        if r1 < 0 or c1 < 0 or c2 > ncols or r1 >= r2 or c1 >= c2:
            raise UserVisibleError(
                i18n.trans(
                    "excel.one_row.badRef",
                    'Excel range "{ref}" is out of bounds',
                    {"ref": range["ref"]},
                )
            )

        # retval of code() is OperatorArray:
        # https://github.com/vinci1it2000/formulas/issues/12
        table_part = list(table.iloc[r1:r2, c1:c2].values.flat)
        formula_args.append(flatten_single_element_lists(table_part))

    # evaluate the formula just once
    # raises ValueError if function isn't implemented
    raw_value = eval_excel(code, formula_args)

    if isinstance(raw_value, Token):
        # XlError('#VALUE!') => '#VALUE!' Text
        return str(raw_value)
    return raw_value


def eval_excel_all_rows(code, table):
    col_idx = []
    for token, obj in code.inputs.items():
        # If the formula is valid but no object comes back it means the
        # reference is no good
        # Missing row number?
        # with only A-Z. But just in case:
        if obj is None:
            raise UserVisibleError(
                i18n.trans(
                    "excel.badCellReference",
                    "Bad cell reference {token}",
                    {"token": token},
                )
            )

        ranges = obj.ranges
        for rng in ranges:
            # r1 and r2 refer to which rows are referenced by the range.
            if rng["r1"] != "1" or rng["r2"] != "1":
                raise UserVisibleError(
                    i18n.trans(
                        "excel.formulaFirstRowReference",
                        "Excel formulas can only reference the first row when applied to all rows",
                    )
                )

            c1 = rng["n1"] - 1
            c2 = rng["n2"]

            if c1 < 0 or c2 > len(table.columns) or c1 >= c2:
                raise UserVisibleError(
                    i18n.trans(
                        "excel.all_rows.badColumnRef",
                        'Excel range "{ref}" is out of bounds',
                        {"ref": rng["ref"]},
                    )
                )

            col_idx.append(list(range(c1, c2)))

    newcol = []
    for row in table.values:
        args_to_excel = [
            flatten_single_element_lists([row[idx] for idx in col]) for col in col_idx
        ]
        # raises ValueError if function isn't implemented
        newcol.append(eval_excel(code, args_to_excel))

    return pd.Series(newcol)


def excel_formula(table, formula, all_rows):
    try:
        # 0 is a list of tokens, 1 is the function builder object
        code = Parser().ast(formula)[1].compile()
    except Exception as e:
        raise UserVisibleError(
            i18n.trans(
                "excel.invalidFormula",
                "Couldn't parse formula: {error}",
                {"error": str(e)},
            )
        )

    if all_rows:
        newcol = eval_excel_all_rows(code, table)
        newcol = autocast_series_dtype(sanitize_series(newcol))
    else:
        # the whole column is blank except first row
        value = eval_excel_one_row(code, table)
        newcol = pd.Series([value] + [None] * (len(table) - 1))

    return newcol


def _get_output_column(table, out_column: str) -> str:
    # if no output column supplied, use result0, result1, etc.
    if not out_column:
        out_column = "result"

    # make sure the colname is unique
    if out_column in table.columns:
        n = 0
        while f"{out_column}{n}" in table.columns:
            n += 1
    else:
        n = ""
    return f"{out_column}{n}"


def _prepare_table_for_excel_formulas(table):
    """Convert columns so they'll work with Excel formulas.

    Excel cannot handle date/timestamp columns; convert those to Excel dates.
    """
    # Extract a table of just the datetimes. They're np.datetime64
    timestamp_table = table[table.columns[table.dtypes == "datetime64[ns]"]]

    # Add date32 columns. They're period[D]; cast them to np.datetime64 because
    # it handles nulls and keeps the math easy. (Float64 has 52-bit fraction:
    # enough to store 32-bit date + 17-bit "86400" seconds/day multiplier, so
    # the conversion is 100% accurate.)
    for colname in table.columns[table.dtypes == "period[D]"]:
        timestamp_table[colname] = table[colname].dt.to_timestamp()
    number_table = pd.DataFrame(
        index=table.index, columns=timestamp_table.columns, dtype=float
    )  # start all-null

    # Excel's number system has two date ranges:
    # 1 .. 60: 1 is 1900-01-01. 60 is 1900-03-01 (1900-02-29 didn't happen)
    # 61 .. *: 61 is 1900-03-01
    # https://docs.microsoft.com/en-gb/office/troubleshoot/excel/wrongly-assumes-1900-is-leap-year
    #
    # But we won't do that. We'll just say 2 is 1900-01-01. This is what
    # LibreOffice and Google Sheets do.
    one_day = np.timedelta64(1, "D").astype("timedelta64[ns]")
    excel_1900_min_date = np.datetime64("1900-01-01").astype("datetime64[ns]")
    excel_1900_zero_date = np.datetime64("1899-12-30").astype("datetime64[ns]")

    number_table[timestamp_table >= excel_1900_min_date] = (
        timestamp_table - excel_1900_zero_date
    ) / one_day
    # Anything else is null

    new_table = table.copy()
    new_table[number_table.columns] = number_table
    return new_table


def render(table, params, **kwargs):
    if table is None:
        return None  # no rows to process

    if params["syntax"] == "excel":
        input_table = _prepare_table_for_excel_formulas(table)
        formula: str = params["formula_excel"]
        if not formula.strip():
            return table
        all_rows: bool = params["all_rows"]
        try:
            newcol = excel_formula(input_table, formula, all_rows)
        except UserVisibleError as e:
            return e.i18n_message
    else:
        formula: str = params["formula_python"].strip()
        if not formula:
            return table
        try:
            newcol = python_formula(table, formula)
        except UserVisibleError as e:
            return e.i18n_message
        except Exception as e:
            return str(e)

    out_column = _get_output_column(table, params["out_column"])
    table[out_column] = newcol

    return table


def _migrate_params_v0_to_v1(params):
    """v0: syntax is int, 0 means excel, 1 means python

    v1: syntax is 'excel' or 'python'
    """
    return {**params, "syntax": ["excel", "python"][params["syntax"]]}


def migrate_params(params):
    if isinstance(params["syntax"], int):
        params = _migrate_params_v0_to_v1(params)
    return params
